src/command_helpers.rs

```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
-        let print = thread::spawn(move || {
-            let mut stderr = BufReader::with_capacity(4096, pipe_reader);
-            let mut line = Vec::with_capacity(20);
-            let stdout = io::stdout();
-
-            // read_until returns 0 on Eof
-            while stderr.read_until(b'\n', &mut line).unwrap() != 0 {
-                {
-                    let mut stdout = stdout.lock();
-
-                    stdout.write_all(b"cargo:warning=").unwrap();
-                    stdout.write_all(&line).unwrap();
-                    stdout.write_all(b"\n").unwrap();
-                }
+const MIN_BUFFER_CAPACITY: usize = 100;
 
-                // read_until does not clear the buffer
-                line.clear();
+impl StderrForwarder {
+    pub(crate) fn new(child: &mut Child) -> Self {
+        Self {
+            inner: child
+                .stderr
+                .take()
+                .map(|stderr| (stderr, Vec::with_capacity(MIN_BUFFER_CAPACITY))),
+            #[cfg(feature = "parallel")]
+            is_non_blocking: false,
+        }
+    }
+
+    fn forward_available(&mut self) -> bool {
+        if let Some((stderr, buffer)) = self.inner.as_mut() {
+            fn write_warning(line: &[u8]) {
+                let stdout = io::stdout();
+                let mut stdout = stdout.lock();
+                stdout.write_all(b"cargo:warning=").unwrap();
+                stdout.write_all(line).unwrap();
+                stdout.write_all(b"\n").unwrap();
             }
-        });
 
-        Ok(Self {
-            handle: Some(print),
-            pipe_writer: Some(pipe_writer),
-        })
+            #[cfg(feature = "parallel")]
+            let read_stderr = if self.is_non_blocking {
+                crate::parallel::stderr::read_non_blocked
+            } else {
+                ChildStderr::read
+            };
+            #[cfg(not(feature = "parallel"))]
+            let read_stderr = ChildStderr::read;
+
+            loop {
+                buffer.reserve(MIN_BUFFER_CAPACITY);
+
+                let old_data_end = buffer.len();
+                buffer.resize(buffer.capacity(), 0);
+                match read_stderr(stderr, &mut buffer[old_data_end..]) {
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        // No data currently, yield back.
+                        buffer.truncate(old_data_end);
+                        return false;
+                    }
+                    Err(err) if err.kind() == std::io::ErrorKind::Interrupted => {
+                        // Interrupted, try again.
+                        buffer.truncate(old_data_end);
+                    }
+                    Ok(0) | Err(_) => {
+                        // End of stream: flush remaining data and bail.
+                        if old_data_end > 0 {
+                            write_warning(&buffer[..old_data_end]);
+                        }
+                        return true;
```

Comment:
Let's set `self.inner = 0;` here, so that once `true` is returned, it will always returns `true` in the future (because in parallel `compile_objects`, it can be called after `true` is returned).

Reply:
Good call.